plugins/vendor/microsoft/teams/batch/KUserGraphMetadata.php

```diff
@@ -0,0 +1,169 @@
+        }
+
+        $this->authToken = $decryptedValue[self::MICROSOFT_GRAPH_BEARER_TOKEN];
+        $this->refreshToken = $decryptedValue[self::MICROSOFT_GRAPH_REFRESH_TOKEN];
+        $this->authTokenExpiry = $decryptedValue[self::MICROSOFT_GRAPH_TOKEN_EXPIRY];
+
+        $this->deltaToken = !empty($metadataObject->xpath('.//DeltaToken')[0]) ? strval($metadataObject->xpath('.//DeltaToken')[0]) : null;
+        $this->lastAccessed = !empty($metadataObject->xpath('.//LastAccessed')[0]) ? intval($metadataObject->xpath('.//LastAccessed')[0]) : null;
+        $this->recordingType = !empty($metadataObject->xpath('.//RecordingType')[0]) ? intval($metadataObject->xpath('.//RecordingType')[0]) : null;
+    }
+
+    public function getXmlFormatted()
+    {
+        $metadataObject = new SimpleXMLElement('<metadata/>');
+        $metadataObject->addChild('GraphAuthData', $this->encryptAuthInfo());
+        $metadataObject->addChild('DeltaToken', $this->deltaToken);
+        $metadataObject->addChild('LastAccessed', $this->lastAccessed);
+        $metadataObject->addChild('RecordingType', $this->recordingType);
+        $metadataObject->addChild('HasOptIn', 'True');
+
+        return $metadataObject->saveXML();
+    }
+
+    private function encryptAuthInfo()
+    {
+        $message = serialize(array(
+            self::MICROSOFT_GRAPH_BEARER_TOKEN => $this->authToken,
+            self::MICROSOFT_GRAPH_REFRESH_TOKEN => $this->refreshToken,
+            self::MICROSOFT_GRAPH_TOKEN_EXPIRY => $this->authTokenExpiry
+        ));
+
+        $iv = openssl_random_pseudo_bytes(16);
+        $value = openssl_encrypt(
+            $message,
+            self::CSRF_CIPHER_METHOD, $this->encryptionKey, 0, $iv
+        );
+        if ($value === false) {
+            throw new Exception('Could not encrypt the data.');
+        }
+        $mac = self::hash($iv = base64_encode($iv), $value, $this->encryptionKey);
+        $json = json_encode(compact('iv', 'value', 'mac'));
+
+        return base64_encode($json);
+    }
+
+    private function decryptAuthInfo($encryptedValue)
+    {
+        $payload = $this->getJsonPayload($encryptedValue, $this->encryptionKey);
+        $iv = base64_decode($payload['iv']);
+        return openssl_decrypt($payload['value'], self::CSRF_CIPHER_METHOD, $this->encryptionKey, 0, $iv);
+    }
+
+    private function getJsonPayload($payload, $key)
+    {
+        $payload = json_decode(base64_decode($payload), true);
+        if (! self::validPayload($payload)) {
+            throw new Exception('The payload is invalid.');
+        }
+        if (! self::validMac($payload, $key)) {
+            throw new Exception('The MAC is invalid.');
+        }
+        return $payload;
+    }
+
+    /**
+     * Calculate the hash of the given payload.
+     *
+     * @param  array  $payload
+     * @param  string  $bytes
+     * @param  string  $key
+     * @return string
+     */
+    private function calculateMac($payload, $bytes, $key)
```

Comment:
some of the functions in this file dont really related to the metadata, and some of them like encrypt/decrypt seems to have almost identical function that exists on kEncryptFileUtils